[Step 1] Bum, JJong

shlim0 opened the pull request:
리뷰어 : @jungseungyeo
팀원 : 범 @snowy-summer, 쫑 @shlim0

---

# 학습목표

- [x]  Swift의 Optional 안전하게 처리하기
    - [Optional Binding](https://developer.apple.com/documentation/swift/optional#Optional-Binding) - guard let을 사용했습니다. if let을 사용하는 대신 guard let을 사용하는 이유는, nil이 아닌 경우 [Early Exit](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/controlflow/#Early-Exit) 되므로 사용자가 정상적인 입력을 하는 경우 더 빠를 것이라 기대했습니다.
- [x]  if와 switch 조건문의 차이와 장단점 비교해보기
    - if의 경우 조건에 일치할 때 까지 제어가 이동하지만, switch의 경우 true가 되는 case가 있으면 바로 이동하게 된다. 따라서, 상대적으로 switch가 더 빠릅니다.
    - if의 경우, 특정 변수값에 대해 조건을 확인할 때 계속 변수를 기입해줘야 하나, switch의 경우 한번 변수를 기입하면 됩니다. 또, loop ({ })를 작성할 필요가 없어 case가 많으면 많을 수록 더욱 편리합니다. 아래 예시를 통해 더 짧고, 간결한 코드 작성이 가능함을 알 수 있습니다.

```swift
// if를 사용한 경우
if value > 100 {
...
} else if value > 200 {
...
} else if value > 300 {
...
}
else {
...
}

// switch를 사용한 경우
switch value {
case ...100:
...
case ...200:
...
case ...300:
...
default:
...
```

- [ ]  함수가 한 가지 일만 하도록 기능 분리하기
- `playRockPaperScissors()`, `checkInput()`, `getGameResult()` 등 기능 별로 함수를 구성했으나 세부 로직을 더욱 간결하게 구성할 필요가 있습니다.
- [ ]  Git의 커밋단위 고민하고 커밋에 적용하기
- 여러 개의 feat, refactor가 추가 됐을 때도 커밋한 경우가 있어 더 고민할 시간이 필요하다 느꼈습니다.
<img width="692" alt="image" src="https://github.com/tasty-code/ios-rock-scissor-paper/assets/46235301/829150bd-077b-4903-b531-b1a416bb376f">

- [ ]  Git 커밋 로그 형식에 맞춰 커밋하기
- [x]  코딩 컨벤션 고민하기
- Swift에서는 함수 뒤에 바로 { 를 붙이지만, 우리 팀에서는 한 라인을 개행하고 { 를 붙이는게 가독성이 좋을 것 같다고 생각했습니다.

# 범

### 코드 설명

- 사용자의 승리 조건을 단순하게 표현하기위해 else if 문의 or조건 으로 표현했던 것을 사용자가 선택한 가위 바위 보에 따른 상황을 딕셔너리로 만들었습니다.
    
    ```swift
    else if input == 1 && computerHand == 3
            || input == 2 && computerHand == 1
            || input == 3 && computerHand == 2
    ```
    
    ```swift
    let winCase : [Int : Int] = [1:3, 2:1, 3:2]
    ```
    

- 사용자의 입력을 받을때 enter키만 입력할 경우 프로그램이 종료되어 종료하지 않고 다시 while문을 실행하는 방법에 대해 고민했습니다.
    - labeled statement을 이용해 다시 while문을 실행하는 방법을 적용했습니다.
    
    ```swift
    RockPaperScissors : while true
    {
        print("가위(1), 바위(2), 보(3)! <종료 : 0> : ", terminator: "")
    
        guard let user = readLine(), let user = Int(user) else
        {
            print("잘못된 입력입니다. 다시 입력해주세요")
            continue RockPaperScissors
        }
    
        switch user
        {
        case 0:
            print("게임 종료")
        case 1...3:
            getGameResult(of: user)
        default:
            print("잘못된 입력입니다. 다시 입력해주세요")
        }
    
        if isExitCondition
        {
            print("----------------------------------------")
            break
        }
    }
    
    ```
    
    - 하지만 동일한 `print("잘못된 입력입니다. 다시 입력해주세요")` 가 반복되어 에러 처리를 위해 enum case를 생성 후 do catch 문을 사용했습니다.
    
    ```swift
    while true
    {
        print("가위(1), 바위(2), 보(3)! <종료 : 0> : ", terminator: "")
    
        do
        {
           try playRockPaperScissors()
        }
        catch RockPaperScissorsError.invalidInput
        {
            print("잘못된 입력입니다. 다시 입력해주세요")
        }
    
        if isExitCondition
        {
            print("----------------------------------------")
            break
        }
    }
    
    ```
    

### 궁금증

- 첫 번째에 선언된 user의 이름을 두 번째에 guard문에서 동일하게 사용 가능한 이유를 모르겠습니다.
    
    ```swift
    guard let user = readLine(), let user = Int(user) else {...}
    ```
    
    ```swift
    guard let user = readLine() else {
    }  // ...1
    
    guard let user = Int(user) else {
    } // ...2
    ```
    
    아래와 같이 guard문 밖에서 user를 사용 가능 한 것으로 알고있는데 왜 서로 다른 범위로 선언된 것인지 모르겠습니다.
    
    ```swift
    guard let user = readLine() else {
    }
    print(user)
    ```
    

---

# 쫑

### 코드 설명

- while문을 두어 가위바위보를 플레이하는 내용을 반복하게 했습니다.
    - 재귀를 하여 반복하는 방법도 있으나 메모리 측면에서 낭비가 된다 판단하여 while문으로 해결했습니다.

- `RockPaperScissorsError` : user 입력이 잘못된 경우 error에 대한 `enum`을 추가하여 예외 처리를 하도록 했습니다.

- `getGameResult()` : 가위바위보 결과에 따라 `flag` 변수를 toggle하였습니다. 또, `isWin()` 을 두어 user가 이기는 경우를 분류하게 했습니다.
    
    

### 궁금증

- while문과 switch문이 혼합되어 있을 때, 제어를 분기하는 방법을 고민했습니다.
    1. `flag 변수`를 전역에 두는 방법
    2. [labeled statement](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/controlflow/#Labeled-Statements)를 활용하는 방법
    - `flag` 변수를 `isExitConditon` 이라 이름지어 분기를 제어했습니다. 그 이유는, `labeled statement`의 경우 C언어에서 `goto`와 유사하기 때문에 스파게티 코드가 될 우려가 있다 생각했기 때문입니다.
    - 앞서 얘기한 두 경우(`flag`,`labeled statement`)를 제외하고 다른 방법이 있는지 궁금합니다.

- 코드를 수정하다보면 commit 단위를 어떻게 설정해야 할지 잘 모르겠습니다.
    - 하나의 commit에 `feat`,`fix`,`refactor` 가 포함되는 경우가 있는데 commit한 내용이 잘 전달되는 것만으로 충분한지 궁금합니다.    

- 추가로, PR 메시지에 작성되어야할 내용과 작성하지 말아야 할 내용이 있다면 알려주시면 좋겠습니다.

### 기타
- 전역 함수를 두지 않고 하나의 struct나 class 내부에서 함수(또는 static 함수)를 구현하여 호출 하는 방식으로 구조 개선을 고려하고 있습니다.

shlim0 committed:
feat : 유저와 컴퓨터의 값을 비교하는 함수 추가

snowy-summer committed:
chore: 기존 2개의 guard문을 1개로 병합

shlim0 committed:
refactor, feat, style, test : 기존 winLoseResult 함수 구조와 네이밍을 변경

refactor :
1. winLoseResult -> getGameResult로 네이밍 변경.
2. input -> user로 parameter 네이밍 변경
4. 함수 내부에서 사용되는 전역 변수를 flag -> isExitCondition으로 네이밍 변경.
5. computerHand -> computer로 네이밍 변경

feat :
1. parameter에 of 접두사 추가
2. else if문에서 유저와 컴퓨터가 낸 패를 비교하는 isWin 딕시너리를 전역에 추가하여 사용

style :
1. if문의 소괄호 () 삭제
2. 기존 flag(변경 후 isExitCondtion) 변수를 전역에서 사용하여 return 불필요하므로 삭제

test :
1. else if문과 else의 디버깅을 위한 print 문 삭제

snowy-summer committed:
refactor, style: 함수의 parameter의 이름에 맞춰 이름 변경

refactor:
1. while문 내부 input -> user 이름 변경

style:
1. while문 { 의 위치 변경

shlim0 committed:
feat, refactor : isWin 함수 추가 및 리팩토링

feat :
1. isWin 함수 추가

rafactor :
1. 전역 딕셔너리 isWin을 함수 isWin으로 옮긴 뒤, winCase로 네이밍 변경
2. isWin 함수 내 딕셔너리 Wincase를 var -> let으로 변경
3. getGameResult 함수 내 기존 isWin 딕셔너리로 승리를 확인하는 것에서 isWin 함수로 승리를 확인하는 것으로 변경

shlim0 committed:
feat : isWin 함수 추가
rafactor :
1. 전역 딕셔너리 isWin을 함수 isWin으로 옮긴 뒤, winCase로 네이밍 변경
2. getGameResult 함수 내 기존 isWin 딕셔너리로 승리를 확인하는 것에서 isWin 함수로 승리를 확인하는 것으로 변경

shlim0 committed:
Merge remote-tracking branch 'origin/JJong' into JJong

snowy-summer committed:
refactor, test, chore : isExitCondition.toggle()의 위치 수정, 코드상 함수 위치 변경, 테스트 코드 삭제

refactor:
1. while문의 종료메시지를 비긴경우가 아니라 승패가 갈렸을 때 나타나도록 변경

test:
1. 테스트 코드 삭제

chore:
1.코드상의 함수의 위치를 변경

shlim0 committed:
Merge remote-tracking branch 'refs/remotes/origin/Bum'

shlim0 committed:
feat, fix : enter만 입력받을 때 생기는 프로그램 종료를 방지

feat :
1. While문에 Labeled Statement인 RockPaperScissors 추가 (46 line)

fix :
1. 유저에게 보여줄 오류 메시지를 추가한 뒤 (52 line) 프로그램이 종료되는 버그를 RockPaperScissors로 continue하는 방식으로 수정 (53 line)

shlim0 committed:
feat, refactor : 에러 처리 및 코드 리팩토링

feat :
1. 에러 메시지를 다룰 열거형 RockPaperScissorsError을 추가

refactor :
1. checkInput 함수를 만들어 while문 내의 기존 guard let으로 입력을 처리하는 코드를 옮김

snowy-summer committed:
style: code convention에 따른 코드 스타일 수정

snowy-summer committed:
feat: while문의 switch문을 playRockPaperScissors 함수로 분리해서 추가

shlim0 committed:
chore : code format을 진행했습니다.

snowy-summer committed:
Merge pull request #1 from snowy-summer/JJong

[Step 1] Bum, JJong

snowy-summer committed:
Update README.md

jungseungyeo commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -2,9 +2,94 @@
+    case invalidInput
+}
+
+while true
```
반복문 while을 쓰셨네요 ㅎㅎ 
혹시 반복문을 while을 선택 한 기준을 알 수 있나요? 

반복문은 while, 재귀,  for문이 있습니다.
다른 반복문 중 무슨 장점으로 선택이 되었을까요?

shlim0 replied:
for를 사용하는 경우 range를 정해줘야 하기 때문에 while 보다 코드가 덜 직관적이라 생각하였습니다.
또, 재귀의 경우 프로그램이 종료될 때 까지 스택 프레임이 쌓이기 때문에 메모리를 낭비하게 될 우려가 있다 판단하였습니다.
따라서, while을 선택하였습니다.

jungseungyeo commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -2,9 +2,94 @@
+func checkInput() throws -> Int
+{
+    guard let user = readLine(), let user = Int(user)
+    else
```
```suggestion
    guard let user = readLine(), let user = Int(user) else {
           throw RockPaperScissorsError.invalidInput
    }
```
위와 같은 형식으로 guard 라인을 바꿔주는 편입니다.

jungseungyeo commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -2,9 +2,94 @@
 import Foundation
 
-print("Hello, World!")
+var isExitCondition: Bool = false
```
이런 형태의 코드를 flag 코드라고해요 개발들 사이에서는 최악의 방법입니다.

최악의 방법은 상태값이 존재하고 어디서 변경이 되는지를 체크 매번 해야 하기 떄문에 
`if isExitCondition` 와 같은 코드 스타일이 매번 들어가게 됩니다. 

이러한 형태가 되지 않는 것이 중요해요. 종료 조건이 되는 함수를 만들고 바로 종료 타이밍이면 끝내는 코드를 수행하면 좋을거가아요.

예를 들어 
shoyldByClose() 이런 함수를 만드어서 종료를 시키는  방법도 있을거같네요 ㅎㅎ
참고만 해주세요

shlim0 replied:
제 생각에도 flag를 두는 것보다 함수에 위임해서 종료시키는 방법이 더 깔끔한 것 같아요 ㅎㅎ
isExitCondition.toggle() 같은 코드가 만들어질 이유도 없어서 좋은 것 같아요.
shouldByClose() 함수를 만들어 프로그램이 종료되게끔 수정했습니다. 다음 번 PR때 확인하실 수 있을 것 같아요.

jungseungyeo commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -2,9 +2,94 @@
+    {
+        case 0:
+            print("게임 종료")
+            isExitCondition.toggle()
```
toggle()은 반복되는 작업을 할 때 좋지만 이 경우 무조건 true가 되어야 되는 경우입니다. 혹시나 휴먼 에러로 두번이 호출된다면 변경이 안되겠죠. 확실히 true를 넣어주는 편이 안전할거같습니다.

jungseungyeo reviewed (approved):
범의 궁금중

`guard를 옵셔널 바인딩을 하기 위해서 사용하셨어요. 1번의 경우 readline()으로 String? -> String으로 옵셔널 바인을 하는 것이고, 두번쨰 경우 string -> Int로 하는 경우입니다.`

위의 경우는 둘다 옵셔널을 하기 때문에 가능합니다.

쫑 질문

`swift goto는 없을 거예요. 이런한 코드 스타일을 말고 좋은 방법은 종료를 하는 역할을 하는 객체 및 함수를 만들고 위임을 주는 것입니다. 에를 들어 지금 끝내도 되니? 와 같은 문구는 if isCloase { break } 이렇게 코드가 작성 되겠지만. 위임을 주어 종료를 시키면 될거같아요.`

feat,fix,refactor와 같이 commit 단위 좋습니다. feat를 기능을 구현하는 단위의 사용을 많이 합니다.

또 고민하시는 struct와 class 구조가 편하시면 언제든지 좋습니다. 저도 struct와 class 구조가 좋습니다.

수고하셨습니다.

shlim0 committed:
feat : 기존 isExitCondition이라는 flag 변수로 프로그램 종료 여부를 체크하는 것에서 shouldByClose()를 추가해 프로그램 종료 여부를 체크하도록 변경함.

shlim0 committed:
chore : shouldByClose() 추가 반영되지 않아 재 push 하였음.

snowy-summer committed:
chore : guard let문 else 문 위치 수정

shlim0 committed:
feat : 새로 만든 RPS.swift에 클래스를 만들어 main.swift에 있던 RockPaperScissors 관련 코드를 이동함.

shlim0 committed:
main.swift
feat :
1. isRepeatRPC() 생성하여 가위바위보를 계속해서 하는 기능을 추가

Refactor :
1. shouldByClose()의 파라미터를 Bool에서 String 타입으로 변경

RPS.swift
1. play()의 return type을 Bool에서 String으로 변경하여 게임 종료시 return "exit" 하게끔 하였음.
2. 1번의 이유로 getResult()의 return type()을 Bool에서 String으로 변경하였음.

snowy-summer committed:
feat: MukJjiPpa 클래스에 누구의 차례인지 확인하는 함수 추가

shlim0 committed:
feat, refactor, chore : RPS 클래스 생성 및 내부 기능 추가

main.swift
feat :
1. MukJjiPpa().isUserTurn() 삭제 후  MukJjiPpa().play() 추가

refactor :
1. RockPaperScissorsError -> ErrorCases로 네이밍 변경
2. RPS.swift 내 isWin()를 main.swift로 이동

chore :
1. isUserTurn 변수 삭제

RPS.swift
chore :
1. getResult() 내부 return 시 "user" -> "사용자", "computer" -> "컴퓨터"로 변경

MukJjiPpa.swift
feat :
1. play() 추가
2. getResult() 추가

refactor :
1. isUserTurn() -> switchTurn()으로 네이밍 변경

snowy-summer committed:
feat, refactor 클래스의 상속 추가, 키워드 변경

feat:
1.MukJjiPpa 클래스가 RPS 클래스를 상속하도록 상속 추가

refactor:
1. repeat 키워드를 draw 로 변경

shlim0 committed:
refactor : 묵찌빠 클래스가 가위바위보 클래스를 상속하는 구조로 변경

main.swift
refactor :
1. gameResult 변수를 RPSResult와 MukJjiPpaResut로 나눔(네이밍 변경). 즉, 각 게임의 결과를 따로 변수를 둠
2. 가위바위보 이후 프로그램 종료를 하지않아도 되므로 isRepeatRPC() 삭제
3. isRepeatRPS() 를 isRepeat로 네이밍 변경
4. isWin()와 checkInput()을 RPS.swift로 이동

RPS.swift
refactor :
1. main.swift에서 while-loop을 도는 구조에서 RPS.swift에서 while-loop을 도는 구조로 변경함. 그에 따라 예외 처리코드도 같이 구성함

MukJjiPpa.swift
refactor :
1. MukJjiPpa 클래스가 RPS 클래스를 상속함
2. play() 내의 whichturn 변수를 private한 인스턴스 프로퍼티로 변경하여 해당 클래스 내부 함수에서 호출할 수 있도록 함
3. getResult 함수를 overriding하여 수정 중

shlim0 committed:
refactor : 가위바위보 코드를 묵찌빠에 맞게 수정

main.swift
refactor :
1. RPSResult 변수를 전역변수로 두어 접근하게 변경

MukJjiPpa.swift
refactor :
1. getResult() 내부 print 내용 및 return 내용 수정
2. isWin()를 '가위바위보' 순서에서 '묵찌빠' 순서에 맞게 수정

shlim0 committed:
fix : 입력 에러시 턴 전환 수정

1. play() 내 switch - default에서 whichTurn = switchTurn(whichTurn)을 whichTurn = "컴퓨터"로 수정함. "컴퓨터"로 입력될 경우 "사용자"로 턴이 변경될 수가 있음.
   (과제 명세 중, 잘못된 입력을 한 경우 컴퓨터에게 턴을 넘기는 조건이 있음)
2. 사용자가 문자로 잘못 입력했을 때, 컴퓨터에게 턴을 넘기지 않는 버그가 있어 checkInput()을 override하여 수정함.

shlim0 committed:
fix :
1. play() 내 switch - default에서 whichTurn = switchTurn(whichTurn)을 whichTurn = "컴퓨터"로 수정함. "컴퓨터"로 입력될 경우 "사용자"로 턴이 변경될 수가 있음.
   (과제 명세 중, 잘못된 입력을 한 경우 컴퓨터에게 턴을 넘기는 조건이 있음)
2. 사용자가 문자로 잘못 입력했을 때, 컴퓨터에게 턴을 넘기지 않는 버그가 있어 checkInput()을 override하여 수정함.

shlim0 committed:
Merge remote-tracking branch 'origin/JJong' into JJong

snowy-summer committed:
chore:  main.swif에 있는 함수들을 각자의 사용하는 클래스에 넣음

isRepeat함수와 shouldByClose함수를 각각의 사용 클래스에 넣음, code convention에 맞게 수정

shlim0 committed:
Merge remote-tracking branch 'refs/remotes/origin/Bum'

Conflicts:
	RockPaperScissors/RockPaperScissors/MukJjiPpa.swift
	RockPaperScissors/RockPaperScissors/RPS.swift
	RockPaperScissors/RockPaperScissors/main.swift

shlim0 committed:
chore : code formatted
